### ISSUE TITLE
Should deprecating a function automatically mark it as "last resort"?

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#17881]
Engin suggested it, but probably only when we make it a user-facing feature:
> I have needed that in different cases. Where one is deprecating an argument name — you have to make the deprecated function a last-resort. Where the other was I believe where you deprecate a non-generic function in favor of a generic one. And if you pass an argument that’s of the type that’s the argument type of the deprecated version, you want that to be a last-resort.

**Comment by bradcray** [comment]
No, I don't think it should.  Marking a function as last resort can sometimes have surprising consequences, such as calling an alternative that relies on promotion or coercion rather than being a better match.  So I think if the goal of deprecation is "keep the program working as before, yet warn people that it won't keep working soon", then I think applying `deprecated` _can't_ result in it being treated as last resort, or the behavior of the program could likely change.

**Comment by e-kayrakli** [comment]
> No, I don't think it should. Marking a function as last resort can sometimes have surprising consequences, such as calling an alternative that relies on promotion or coercion rather than being a better match. So I think if the goal of deprecation is "keep the program working as before, yet warn people that it won't keep working soon", then I think applying deprecated can't result in it being treated as last resort, or the behavior of the program could likely change.

This is reasonable.

> [edit: That said, I personally find "last resort" a sufficiently compelling feature that I think we should provide it to end-users in some form as well]

This is also interesting.

Lydia already quoted me on this, but I just want flesh out a case where a user-facing deprecation feature can fall short without `last resort` or `deprecated` (or equivalent) being able to break the tie between two resolution candidates:

```chapel
deprecated "Formal argument name 'x' is deprecated use 'y' instead"
proc foo(x) {

}

proc foo(y) {

}

foo(5);
```

without making the deprecated version a `last resort`, this code is ambiguous.

PS. If you make it a `last resort` today, it suppresses the deprecation warning. Which is not really surprising.

**Comment by bradcray** [comment]
> If you make it a last resort today, it suppresses the deprecation warning. Which is not really surprising.

Just in case anyone misinterprets this as I originally did:  The call `foo(5)` above wouldn't generate the deprecation warning because it would prefer the non-deprecated version.  But calling `foo(x=5)` does still generate the deprecation warning, as you'd hope/expect.
